# nf-core/taxprofiler: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. *Nat Biotechnol.* 2020;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. *Nat Biotechnol.* 2017;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools

- [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

- [MultiQC](https://pubmed.ncbi.nlm.nih.gov/27312411/)

  > Ewels P, Magnusson M, Lundin S, Käller M. MultiQC: summarize analysis results for multiple tools and samples in a single report. *Bioinformatics.* 2016;32(19):3047-8. doi: 10.1093/bioinformatics/btw354. Epub 2016 Jun 16. PubMed PMID: 27312411; PubMed Central PMCID: PMC5039924.

- [falco](https://doi.org/10.12688/f1000research.21142.2)

  > de Sena Brandine G & Smith AD. Falco: high-speed FastQC emulation for quality control of sequencing data. *F1000Research.* 2021; 8:1874.

- [fastp](https://doi.org/10.1093/bioinformatics/bty560)

  > Chen S, Zhou Y, Chen Y, Gu J. Fastp: An Ultra-Fast All-in-One FASTQ Preprocessor. *Bioinformatics.* 2018;34(17):i884-90. doi: 10.1093/bioinformatics/bty560.

- [AdapterRemoval2](https://doi.org/10.1186/s13104-016-1900-2)

  > Schubert M, Lindgreen S, and Orlando L. AdapterRemoval v2: Rapid Adapter Trimming, Identification, and Read Merging. *BMC Research Notes.* 2016;9:88. doi:10.1186/s13104-016-1900-2.

- [Porechop](https://github.com/rrwick/Porechop)

- [FILTLONG](https://github.com/rrwick/Filtlong)

- [BBTools](http://sourceforge.net/projects/bbmap/)

- [PRINSEQ++](https://doi.org/10.7287/peerj.preprints.27553v1)

  > Cantu VA, Sadural J, and Edwards R. PRINSEQ++, a Multi-Threaded Tool for Fast and Efficient Quality Control and Preprocessing of Sequencing Datasets. *PeerJ Preprints.* 2019;e27553v1. doi: 10.7287/peerj.preprints.27553v1.

- [Bowtie2](https://doi.org/10.1038/nmeth.1923)

  > Langmead B & Salzberg SL. Fast gapped-read alignment with Bowtie 2. *Nature Methods.* 2012;9(4):357–359. doi: 10.1038/nmeth.1923

- [minimap2](https://doi.org/10.1093/bioinformatics/bty191)

  > Li H. Minimap2: pairwise alignment for nucleotide sequences. *Bioinformatics.* 2018;34(18):3094–3100. doi: 10.1093/bioinformatics/bty191

- [SAMTools](https://doi.org/10.1093/gigascience/giab008)

  > Danecek P, Bonfield JK, Liddle J, et al. Twelve years of SAMtools and BCFtools. *GigaScience.* 2021;10(2). doi: 10.1093/gigascience/giab008

- [MetaPhlAn4](https://doi.org/10.7554/eLife.65088)

  > Beghini F, McIver L, Blanco-Míguez A, et al. Integrating Taxonomic, Functional, and Strain-Level Profiling of Diverse Microbial Communities with BioBakery 3. *ELife.* 2021;10:e65088. doi: 10.7554/eLife.65088

- [Sourmash](https://doi.org/10.12688/f1000research.19675.1)

  > Pierce NT, Irber L, Reiter T, Brooks P, Brown CT. Large-scale sequence comparisons with sourmash. *F1000Research.* 2019;8:1006. doi: 10.12688/f1000research.19675.1

- [bwa - Burrows-Wheeler Alignment Tool](https://bio-bwa.sourceforge.net/bwa.shtml)

- [AMRplusplus](https://doi.org/10.1093/nar/gkac1047)

  > Bonin N, Doster E, Worley H, et al. MEGARes and AMR++, v3.0: an updated comprehensive database of antimicrobial resistance determinants and an improved software pipeline for classification using high-throughput sequencing. *Nucleic Acids Research.* 2023;51(1):744-52. doi: 10.1093/nar/gkac1047

- [QIIME2](https://doi.org/10.1038%2Fs41587-019-0209-9)

  > Bolyen E, Rideout JR, Dillon MR, et al. Reproducible, interactive, scalable, and extensible microbiome data science using QIIME 2. *Nat Biotechnol.* 2020;37(8):852-857. doi: 10.1038/s41587-019-0209-9

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)

  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)

  > Grüning B, Dale R, Sjödin A, et al. Bioconda: sustainable and comprehensive software distribution for the life sciences. *Nat Methods.* 2018;15(7):475-476. doi: 10.1038/s41592-018-0046-7. PubMed PMID: 29967506.

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)

  > da Veiga Leprevost F, Grüning B, et al. BioContainers: an open-source and community-driven framework for software standardization. *Bioinformatics.* 2017;33(16):2580-2582. doi: 10.1093/bioinformatics/btx192. PubMed PMID: 28379341; PubMed Central PMCID: PMC5870671.

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)
  > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. *PLoS One.* 2017;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.

## Data

- [Maixner (2021)](https://doi.org/10.1016/j.cub.2021.09.031) (CI Test Data)

  > Maixner F, Sarhan M, Huang K, et al. Hallstatt Miners Consumed Blue Cheese and Beer during the Iron Age and Retained a Non-Westernized Gut Microbiome until the Baroque Period. *Current Biology.* 2021;31(23): 5149–62.e6. doi: 10.1016/j.cub.2021.09.031.

- [Meslier (2022)](https://doi.org/10.1038/s41597-022-01762-z) (AWS Full Test data)

  > Meslier V, Quinquis B, Da Silva K, et al. Benchmarking Second and Third-Generation Sequencing Platforms for Microbial Metagenomics. *Scientific Data.* 2022;9(1):694. doi: 10.1038/s41597-022-01762-z.
